src/main/java/io/cryostat/graphql/EnvironmentNodes.java

```diff
@@ -0,0 +1,133 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.eclipse.microprofile.graphql.Description;
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Query;
+
+import io.cryostat.discovery.DiscoveryNode;
+import io.smallrye.graphql.api.Nullable;
+
+@GraphQLApi
+public class EnvironmentNodes {
+
+    public static class EnvironmentNodeFilterInput {
+        @Nullable
+        public Long id;
+        @Nullable
+        public String name;
+        @Nullable
+        public List<String> names;
+        @Nullable
+        public String nodeType;
+        @Nullable
+        public List<String> labels;
+
+    }
+
+    public static class EnvironmentNode {
+        private Long id;
+        private String name;
+        private List<String> labels;
+        private String nodeType;
+        private List<EnvironmentNode> children;
+
+        public Long getId(){
+            return this.id;
+        }
+
+        public String getName(){
+            return this.name;
+        }
+
+        public List<String> getLabels(){
+            return this.labels;
+        }
+
+        public String getNodeType(){
+            return this.nodeType;
+        }
+
+        public List<EnvironmentNode> getChildren(){
+            return this.children;
+        }
+
+        public void setId(Long id) {
+            this.id = id;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public void setLabels(List<String> labels) {
+            this.labels = labels;
+        }
+
+        public void setNodeType(String nodeType) {
+            this.nodeType = nodeType;
+        }
+
+        public void addChild(EnvironmentNode child) {
+            if (this.children == null) {
+                this.children = new ArrayList<>();
+            }
+            this.children.add(child);
+        }
+    }
+
+    @Query("environmentNodes")
+    @Description("Get all environment nodes in the discovery tree with optional filtering")
+    public List<EnvironmentNode> environmentNodes(EnvironmentNodeFilterInput filter) {
+        DiscoveryNode rootNode = DiscoveryNode.getUniverse();
+        return filterAndTraverse(rootNode, filter).stream()
+            .map(EnvironmentNodes::convertDiscoveryNodeToEnvironmentNode)
+            .collect(Collectors.toList());
+    }
+
+    private List<DiscoveryNode> filterAndTraverse(DiscoveryNode node, EnvironmentNodeFilterInput filter) {
+        List<DiscoveryNode> filteredNodes = new ArrayList<>();
+        if (matchesFilter(node, filter)) {
+            filteredNodes.add(node);
+        }
+        if (node.children != null) {
+            for (DiscoveryNode child : node.children) {
+                filteredNodes.addAll(filterAndTraverse(child, filter));
+            }
+        }
+        return filteredNodes;
+    }
+
+    private static boolean matchesFilter(DiscoveryNode node, EnvironmentNodeFilterInput filter) {
```

Comment:
Part of the filter here should be to check the [`discoveryNode.target == null`](https://github.com/cryostatio/cryostat3/blob/f64d7656ceaae57eb6ac03e017d30f67f997bc1c/src/main/java/io/cryostat/discovery/DiscoveryNode.java#L90). In the new data model, that's probably the best way to distinguish "environment nodes" from "target nodes".